compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -8185,4 +8166,295 @@ private boolean isEndOfTypeList(SyntaxKind nextTokenKind) {
+        STNode queryPipeline = parseQueryPipeline();
+        STNode selectClause = parseSelectClause();;
+        endContext();
+        return STNodeFactory.createQueryExpressionNode(queryConstructType, queryPipeline, selectClause);
+    }
+
+    /**
+     * Parse query construct type.
+     * <p>
+     * <code>query-construct-type := table key-specifier</code>
+     *
+     * @return Parsed node
+     */
+    private STNode parseQueryConstructType(STNode tableKeyword, STNode keySpecifier) {
+        return STNodeFactory.createQueryConstructTypeNode(tableKeyword, keySpecifier);
+    }
+
+    /**
+     * Parse query pipeline.
+     * <p>
+     * <code>
+     * query-pipeline := from-clause intermediate-clause*
+     * <br/>
+     * intermediate-clause := from-clause | where-clause | let-clause
+     * </code>
+     *
+     * @return Parsed node
+     */
+    private STNode parseQueryPipeline() {
+        STNode fromClause = parseFromClause();
+        List<STNode> clauses = new ArrayList<>();
+        return parseQueryPipeline(fromClause, clauses);
+    }
+
+    private STNode parseQueryPipeline(STNode fromClause, List<STNode> clauses) {
+        return parseQueryPipeline(peek().kind, fromClause, clauses);
+    }
+    private STNode parseQueryPipeline(SyntaxKind nextTokenKind, STNode fromClause, List<STNode> clauses) {
+        STNode clause;
+
+        while (true) {
+            switch (nextTokenKind) {
+                case FROM_KEYWORD:
+                    clause = parseFromClause();
+                    clauses.add(clause);
+                    nextTokenKind = peek().kind;
+                    continue;
+                case WHERE_KEYWORD:
+                    clause = parseWhereClause();
+                    clauses.add(clause);
+                    nextTokenKind = peek().kind;
+                    continue;
+                case LET_KEYWORD:
+                    clause = parseLetClause();
+                    clauses.add(clause);
+                    nextTokenKind = peek().kind;
+                    continue;
+                case SELECT_KEYWORD:
+                    break;
+                default:
+                    if (nextTokenKind == SyntaxKind.EOF_TOKEN) {
+                        return parseQueryPipeline(SyntaxKind.SELECT_KEYWORD, fromClause, clauses);
+                    }
+
+                    Solution solution = recover(peek(), ParserRuleContext.QUERY_EXPRESSION_RHS, fromClause, clauses);
+
+                    // If the parser recovered by inserting a token, then try to re-parse the same
+                    // rule with the inserted token. This is done to pick the correct branch
+                    // to continue the parsing.
+                    if (solution.action == Action.REMOVE) {
+                        return solution.recoveredNode;
+                    }
+
+                    return parseQueryPipeline(solution.tokenKind, fromClause, clauses);
+            }
```

Comment:
We can extract the content inside the while loop to a method. That method simply parse and returns a clause. We have done a similar thing in the `parseFollowUpArg()` method.

Reply:
Suppose we extract this to a new new method say parseIntermediateClause. Then what should be return if Solution solution = recover(),  recovered a SELECT_KEYWORD? We can't pass select clause inside that method as it should be passed after the parseQueryPipeline

Reply:
we can actually parse is as select-clause. caller can check the kind of the clause returned by this method and stop the loop, is its a select clause.

Reply:
+1

Reply:
``` java 
private STNode parseQueryExprRhs(STNode queryConstructType) {
        switchContext(ParserRuleContext.QUERY_EXPRESSION);
        STNode fromClause = parseFromClause();

        List<STNode> clauses = new ArrayList<>();
        STNode clause = null;

        while (!isEndOfIntermediateClause(peek().kind)) {
            clause = parseIntermediateClause();
            if (clause.kind == SyntaxKind.SELECT_CLAUSE) {
                break;
            }
            clauses.add(clause);
        }

        if (clause == null || clause.kind != SyntaxKind.SELECT_CLAUSE) {
            clause = parseSelectClause();
        }
        
        STNode intermediateClauses = STNodeFactory.createNodeList(clauses);
        STNode queryPipeline = STNodeFactory.createQueryPipelineNode(fromClause, intermediateClauses);
        return STNodeFactory.createQueryExpressionNode(queryConstructType, queryPipeline, clause);
    }
```

Reply:
```java    
private STNode parseQueryExprRhs(STNode queryConstructType) {
        switchContext(ParserRuleContext.QUERY_EXPRESSION);
        STNode fromClause = parseFromClause();

        STNode clause;
        STNode intermediateClauses;
        STNode queryPipeline;

        List<STNode> clauses = new ArrayList<>();
        while (!isEndOfIntermediateClause(peek().kind)) {
            clause = parseIntermediateClause();

            if (clause.kind == SyntaxKind.SELECT_CLAUSE) {
                intermediateClauses = STNodeFactory.createNodeList(clauses);
                queryPipeline = STNodeFactory.createQueryPipelineNode(fromClause, intermediateClauses);
                return STNodeFactory.createQueryExpressionNode(queryConstructType, queryPipeline, clause);
            }
            
            clauses.add(clause);
        }

        clause = parseSelectClause();
        intermediateClauses = STNodeFactory.createNodeList(clauses);
        queryPipeline = STNodeFactory.createQueryPipelineNode(fromClause, intermediateClauses);
        return STNodeFactory.createQueryExpressionNode(queryConstructType, queryPipeline, clause);
    }
```

Reply:
two approaches. Any suggestion or improvement?